src/Cardano/Shell/Update/Types.hs

```diff
@@ -0,0 +1,114 @@
+module Cardano.Shell.Update.Types where
+
+import           Cardano.Prelude
+
+import qualified Data.Map as M
+
+import           Test.QuickCheck
+
+
+-- Types we need to check the behaviour.
+data Blockchain = Blockchain { getBlockchainContents :: !(Map Epoch [Slot]) }
+    deriving (Eq, Show)
+
+data Epoch = Epoch Int
+    deriving (Eq, Ord, Show)
+
+data Slot = Slot (Maybe Block)
+    deriving (Eq, Show)
+
+data Block = Block BlockHash (Maybe InstallerVersion)
+    deriving (Eq, Show)
+
+data BlockHash = BlockHash !Text
+    deriving (Eq, Show)
+
+data InstallerVersion = InstallerVersion InstallerHash
+    deriving (Eq, Ord, Show)
+
+data InstallerHash = InstallerHash !Text
+    deriving (Eq, Ord, Show)
+
+-- | The wrapper is used since this is an important distinction.
+newtype LatestInstallerVersion = LatestInstallerVersion
+    { getInstallerVersion :: InstallerVersion
+    } deriving (Eq, Show)
+
+-- | Function for fetching @InstallerVersion@ from a @Block@ from a @Slot@.
+slotContainsInstaller :: Slot -> Maybe InstallerVersion
+slotContainsInstaller (Slot Nothing)    = Nothing
+slotContainsInstaller (Slot block)      = blockContainsInstaller =<< block
+  where
+    blockContainsInstaller :: Block -> Maybe InstallerVersion
+    blockContainsInstaller (Block _ Nothing)          = Nothing
+    blockContainsInstaller (Block _ installerVersion) = installerVersion
+
+
+--------------------------------------------------------------------------------
+-- Generation functions
+--------------------------------------------------------------------------------
+
+genInstallerVersion :: Gen InstallerVersion
+genInstallerVersion = InstallerVersion <$> genInstallerHash
+
+genInstallerHash :: Gen InstallerHash
+genInstallerHash = InstallerHash <$> genSafeText
+
+-- | Generate random ascii string, it's very simplifed.
+genSafeText :: Gen Text
+genSafeText = strConv Lenient <$> (listOf1 $ choose ('a', 'z'))
+
+genBlock :: Gen Block
+genBlock = Block <$> genBlockHash <*> genMaybeInstallerVersion
+
+genSlot :: Gen Slot
+genSlot = Slot <$> frequency
+    [ (10, Just <$> genBlock)
+    , (1 , pure Nothing)
+    ]
+
+genSlots :: Gen [Slot]
+genSlots = replicateM 10 genSlot
+
+genBlockHash :: Gen BlockHash
+genBlockHash = BlockHash <$> genSafeText
+
+genMaybeInstallerVersion :: Gen (Maybe InstallerVersion)
+genMaybeInstallerVersion = frequency
+    [ (1 , Just <$> genInstallerVersion)
+    , (10, pure Nothing)
+    ]
+
+-- | Here we generate the blockchain that contains the installer
+genBlockchainWithInstaller :: Gen Blockchain
+genBlockchainWithInstaller = do
+    numOfEpochs <- choose (1, 100)
+
+    let availableEpochs = map Epoch [1..numOfEpochs]
```

Comment:
You can use `coerce` if `Epoch` is defined with `newtype`

Reply:
Do NOT use coerce. It makes things unclear and less explicit.